server/handlers.go

```diff
@@ -1319,21 +1326,112 @@ func (s *Server) handlePasswordGrant(w http.ResponseWriter, r *http.Request, cli
 	s.writeAccessToken(w, resp)
 }
 
+func (s *Server) handleTokenExchange(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// TODO: check global allowed grant types?
+
+	if err := r.ParseForm(); err != nil {
+		s.logger.Errorf("could not parse request body: %v", err)
+		s.tokenErrHelper(w, errInvalidRequest, "", http.StatusBadRequest)
+		return
+	}
+	q := r.Form
+
+	resource := q.Get("resource")                       // OPTIONAL, use for issued token audience
```

Comment:
I started thinking that adding this parameter was not at all right.
In Dex, the audience claim always points to a client (or clients). There may be private or public clients.

For private clients, credentials are always required. With the `resource` parameter now, users can issue tokens for any client they want without credentials (even for not existing ones, which seems to be a hole.

Before merging, we need to resolve this. One way is to use the client authentication, as for some other grants (with a decorator), and abstain from the resource parameter. https://github.com/dexidp/dex/blob/2e4d724935fb121aff49a0bbf35325eafc79fa57/server/handlers.go#L841

The other option is to reject any requests with the resource parameter equals to one of the existing client ids (which still leaves space for security holes but is less problematic).

Reply:
It is also strange because the audience claim is required, but the resource parameter is optional. Thus, by default, each token has an empty aud claim (because the aud claim is required by spec).

I checked some popular providers:
1. Okta uses the audience parameter value for the audience claim value https://developer.okta.com/docs/guides/set-up-token-exchange/main/#token-exchange-request-from-service-app-to-api
2. Keycloak uses the audience claim as the client_id, but they also have semi-required client id and client secret parameters (as suggested per my top comment) https://www.keycloak.org/docs/latest/securing_apps/#_token-exchange

We probably need to use our usual `connector` parameter to specify the connector instead of the resource. What do you think?

Reply:
I hadn't considered that dex required the issued token audience to be that of a known client (I've been working with sigstore/cosign in gcp recently, sigstore requires a specific audience and gcp doesn't require registering clients).

If I understand correctly, with dex's support cross client trust, we need 3 pieces of information:
- current client id
  - we can, and should require client auth via the decorator like with other handlers
- target client id (optional)
  - use `audience` in the incoming request, defaults to current client id
- connector id
  - use `connector_id` like the `/auth` endpoint currently does
  - I thought about using `resource` for this, but rereading the RFC it seems `resource` is closer to `audience`

Thoughts on this approach?

Reply:
actually, no dex uses a scope for cross client trust (target client id).
Then I think we should just ignore `resource` and `audience` in the incoming request then

Reply:
I've updated the PR to match the approach I laid out above.